keps/173-headless-service-per-replica/README.md

```diff
@@ -0,0 +1,328 @@
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### API
+
+```golang
+type LeaderWorkerSetSpec struct {
+ // SubdomainPolicy determines the policy that will be used when creating
+ // the headless service
+ SubdomainPolicy SubdomainPolicy `json:"subdomainPolicy,omitempty"`
+}
+
+type NetworkConfig struct {
+  SubdomainPolicy SubdomainPolicy
+}
+type SubdomainPolicy string
+const (
+  // SubdomainShared will create a single headless service that all replicas 
+  // will share
+  SubdomainShared SubdomainPolicy = "Shared"
+  // SubdomainUniquePerReplica will create a headless service for each
+  // leader-worker group
+  SubdomainUniquePerReplica SubdomainPolicy = "UniquePerReplica"
+)
+```
+
+### Implementation
+
+With HeadlessServicePerReplica true, we will create a headless service per replica.
+The name of the service will be `{lws.Name}-{replicaIndex}`.
+Creation of the services will be in a loop of replicas. The selector of each
+
+With SubdomainPolicy set to SubdomainUniquePerReplica, we will create a headless service
+per replica. The name of the service will be `{lws.Name}-{replicaIndex}`. The selector of
+the headless service will be the group index.
+
+```golang
+for i := 0; i < lws.Spec.Replicas; i++ {
+  headlessService := corev1.Service{
+    ObjectMeta: metav1.ObjectMeta{
+      Name:      fmt.Sprintf("%s-%s", lws.Name, i),
+      Namespace: lws.Namespace,
+    },
+    Spec: corev1.ServiceSpec{
+      ClusterIP: "None", // defines service as headless
+      Selector: map[string]string{
+        leaderworkerset.GroupIndexLabelKey: i,
+      },
+      PublishNotReadyAddresses: true,
+    },
+  }
+}
+```
+
+Currently, LWS will add an environment variable to all pods that
+points to the dns record.
+
+The pods get the following environment variable:
+
+```golang
+ leaderAddressEnvVar := corev1.EnvVar{
+  Name:  leaderworkerset.LwsLeaderAddress,
+  Value: fmt.Sprintf("%s-%s.%s.%s", lwsName, groupIndex, lwsName, pod.ObjectMeta.Namespace),
+ }
+```
+
+With this feature enabled, we will need to set the headless service of the replica the pod belongs to.
+```golang
+ leaderAddressEnvVar := corev1.EnvVar{
+  Name:  leaderworkerset.LwsLeaderAddress,
+  Value: fmt.Sprintf("%s-%s.%s-%s.%s", lwsName, groupIndex, lwsName, groupIndex, pod.ObjectMeta.Namespace),
+ }
+```
+
```

Comment:
we may need to add implementation details for validation/update webhook logic. Also do we allow mutating this field etc?

Reply:
+1, if we want to support updates, that would be great, but we need to clearly discuss how we will do that without disruptions.

Reply:
Added details for update logic. After testing, I think it is doable to allow mutating the field.